api/payara-api/src/main/java/fish/payara/context/ContextProducer.java

```diff
@@ -0,0 +1,171 @@
+/*
+ *  DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ *  Copyright (c) [2020] Payara Foundation and/or its affiliates. All rights reserved.
+ *
+ *  The contents of this file are subject to the terms of either the GNU
+ *  General Public License Version 2 only ("GPL") or the Common Development
+ *  and Distribution License("CDDL") (collectively, the "License").  You
+ *  may not use this file except in compliance with the License.  You can
+ *  obtain a copy of the License at
+ *  https://github.com/payara/Payara/blob/master/LICENSE.txt
+ *  See the License for the specific
+ *  language governing permissions and limitations under the License.
+ *
+ *  When distributing the software, include this License Header Notice in each
+ *  file and include the License file at glassfish/legal/LICENSE.txt.
+ *
+ *  GPL Classpath Exception:
+ *  The Payara Foundation designates this particular file as subject to the "Classpath"
+ *  exception as provided by the Payara Foundation in the GPL Version 2 section of the License
+ *  file that accompanied this code.
+ *
+ *  Modifications:
+ *  If applicable, add the following below the License Header, with the fields
+ *  enclosed by brackets [] replaced by your own identifying information:
+ *  "Portions Copyright [year] [name of copyright owner]"
+ *
+ *  Contributor(s):
+ *  If you wish your version of this file to be governed by only the CDDL or
+ *  only the GPL Version 2, indicate your decision by adding "[Contributor]
+ *  elects to include this software in this distribution under the [CDDL or GPL
+ *  Version 2] license."  If you don't indicate a single choice of license, a
+ *  recipient has the option to distribute your version of this file under
+ *  either the CDDL, the GPL Version 2 or to extend the choice of license to
+ *  its licensees as provided above.  However, if you add GPL Version 2 code
+ *  and therefore, elected the GPL Version 2 license, then the option applies
+ *  only if the new code is made subject to such option by the copyright
+ *  holder.
+ */
+package fish.payara.context;
+
+import java.io.Serializable;
+
+/**
+ * Utility to create / push Jakarta EE and CDI thread contexts
+ *
+ * Example:
+ * @Inject ContextProducer producer;
+ * // EJB/CDI thread:
+ * Instance saved = producer.currentInvocation*();
+ * // insure 'saved' is not leaked when application undeployed,
+ * // otherwise use producer.fromComponentId(producer.getInvocationComponentId())
+ * // and in another, non EJB/CDI thread:
+ * try (Context ctx = saved.pushRequestContext()) {
+ *     // runs with EJB / CDI context
+ * }
+ *
+ * @author lprimak
+ */
+public interface ContextProducer {
```

Comment:
Is this really something we encourage users to use in their applications, and promise to keep API binary compatible?

I think this should have home elsewhere, somewhere under `nucleus/common`.

Reply:
nucleus/common is where I started with this before migrating to Payara-api.
This was designed before MP context propagation, and perhaps we should use that in the future,
but it's not a reason to hold up this PR currently.

Reply:
There's also Java EE Concurrency API, that deals with context propagation. But neither of these APIs expose internals of application servers, like invocation stack or underlying classloaders and give explicit control over them to user application. They only allow propagating current context to async calls.

I don't believe this belongs into application server's public api.

Reply:
Do you have an alternative? How else is Hazelcast callback supposed to deal with EJB/CDI objects?
I don't look at it as 'App server internals' rather than an integration point that let's apps use Jakarta EE.

Other frameworks always touted 'we have a main method' type approach, this gives and entry-way to do something like that on a hybrid basis at least, which is very useful

Reply:
I cannot offer an alternative, because I don't know what the exact requirements are. Does any of the other PRs demonstrate how would user app use these?

Jakarta EE is all about integration, yet so far they seemed to solve this without giving fine-grained control over invocation context to user. I can give you lots of examples that are probably not relevant, but may help understanding the motivation. And then we can look at whether the API could be more opaque, and then look at security implications of this.

Is there anything in JCA's inbound flows (so special kind of message driven bean) that would cover these use cases? Could CDI events be produced by application server instead of user app directly interacting with hazelcast over callback? An interceptor that would get installed around Hazelcast callback by the application server and manage the stack?

Reply:
Ah I now saw the example in the PR description, I can see few alternatives to write it:

Reply:
I edited the top of the PR with an example.
You do raise a good point about security, as this API makes it much easier to switch context into another application running on the same server (if ran without security manager)

Reply:
I found out, that I'm not entirely sure what the code sample is trying to achive.
```

        map.put("hello", "world"); // (1)

        map2.put("bye", listener); // (2)
```
What I think that (1) does is, that when when entry is added to map, it will invoke `@Inject MyListener listener` asynchronoulsy in current context. i.e invocation of Singleton EJB.

For this I can imagine
- Using `javax.enterprise.concurrent.ContextService`, its api is strinkingly similar.
- Set of Hazelcast specific APIs, i. e. CDI extension that would install map listeners and then invoke event observers (`void entryAdded(@Observes @EntryAdded(mapName = "hallo") String entry) {}`)


I'm not sure what (2) is trying to achieve. What should happen when we put managed object into distributed map? How do we manage it serializes properly since it the reference might be just a proxy (to a proxy) of the application class? How do we guarantee that map is updated when objects (or proxies) internal state changes? How do we cope with tha fact, that the grid is usually not running single version of an application and therefore serialization compatibility of the managed object must be retained? What if effects of invoking a method on the objects are local only?

Reply:
Jakarta EE has 'solved' - not really? this problem by mandating that objects can be used only in App server control flow. If we want true Hazelcast integration (which people are asking for), some sort of context API needs to be exposed.

Also, keep this in mind: https://github.com/payara/Payara/pull/5013
I am also working on tenant control for Hazelcast which will make it possible for some Hazelcast objects to automatically propagate Payara app context, but some cannot and still need an API to do this

Reply:
> Set of Hazelcast specific APIs, i. e. CDI extension that would install map listeners and then invoke event observers (void entryAdded(@Observes @EntryAdded(mapName = "hallo") String entry) {})

This is clearly out of scope for this PR

> Using javax.enterprise.concurrent.ContextService, its api is strinkingly similar.

Again, this was pre-MP context propagation, and if we are going to use something, it's MP stuff I would think
I thin it would also be out of scope here